state-network.md

```diff
@@ -125,54 +86,107 @@ A node should track their own radius value and provide this value in all Ping or
+
+
 #### Component Data Elements
 
-#### Proofs
+#### Merkle Patricia Trie (MPT) Proofs
+
 Merkle Patricia Trie (MPT) proofs consist of a list of witness nodes that correspond to each trie node that consists of various data elements depending on the type of node (e.g.blank, branch, extension, leaf).  When serialized, each witness node is represented as an RLP serialized list of the component elements with the largest possible node type being the branch node which when serialized is a list of up to sixteen hashes in `Bytes32` (representing the hashes of each of the 16 nodes in that branch and level of the tree) plus the 4 elements of the node's value (balance, nonce, codehash, storageroot) represented as `Bytes32`.  When combined with the RLP prefixes, this yields a possible maximum length of 667 bytes.  We specify 1024 as the maximum length due to constraints in the SSZ spec for list lengths being a power of 2 (for easier merkleization.)
+
+> TODO: Define proof such that there is one canonical representation of a witness that is streamable and minimal??
+> Specs should specify order of proof nodes, and disallow inclusion of superflous nodes.
+
 ```
 WitnessNode            := ByteList(1024)
 MPTWitness             := List(witness: WitnessNode, max_length=32)
 ```
 
-#### Account Trie Proof
+#### Paths (Nibbles)
 
-A leaf node from the main account trie and accompanying merkle proof against a recent `Header.state_root`
+A naive approach to storage of trie nodes would be to simply use the `node_hash` value of the trie node for storage.  This scheme however results in stored data not being tied in any direct way to it's location in the trie.  In a situation where a participant in the DHT wished to re-gossip data that they have stored, they would need to reconstruct a valid trie proof for that data in order to construct the appropriate OFFER/ACCEPT payload.  We include the `path` metadata in state network content keys so that it is possible to reconstruct this proof.
 
+We define path as a sequences of "nibbles" which represent the path through the MPT to reach the trie node.
+
+```
+nibble := {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, a, b, c, d, e, f} // 16 possible values
+NibblePair := Byte // 2 nibbles tightly packed into a single byte
+Nibbles := List(NibblePair, max_length=32) // fixed path length of 8 bytes
 ```
-account_trie_proof_key := Container(address: Bytes20, state_root: Bytes32)
+
+
+#### Helper Functions
+
+We define these helper functions.
+
+##### `encode_path(path: Nibbles) -> ???`
+
+> TODO: what is the most efficient way to encode nibbles and place them in an SSZ container...
+
+- for state roots, the path is the empty list.
+- for intermediate nodes we expect between 1-9 nibbles values, aka roughly 3-6 bytes when tightly packed and using terminator byte
+- for leaf nodes we can use the 20-byte address pre-image to compress the actual 32-byte path in the trie.
```

Comment:
This works only for account tree, but not for the storage tree.

The leaf nodes already contain nibble prefix since the previous node, so I believe that we can just encode the path up to the previous node (I think it has to be branch) + 1 nibble. We can use the same logic for the extension nodes.

Reply:
I don't think this works since the trie around the leaf node could change, increasing the node's depth in the trie.

Basically, anything doing parent referencing seems problematic for this reason...

I think just including the full 32 byte path for contract storage nodes (or maybe the "slot" pre-image since this might save space)

Reply:
This appears to be correct (I had to go read through the py-trie implementation for a bit to remember the subtleties of how this works).

I think your suggestion is the right way forward for probably both the account and storage trie leaves.

I'm also realizing that we may not actually have access to the "pre-images" since that would require the client generating the proofs to give those to us which it may not be tooled correctly for.